Runtime/ui/painting/txt/text_blob.cs

```diff
@@ -1,69 +1,97 @@
-
-        Rect _boundsInText;
-    }
-
-    public class TextBlobBuilder {
-        TextStyle _style;
-        public Vector2d[] positions;
-        string _text;
-        int _textOffset;
-        int _size;
-        Rect _bounds;
-
-        public void allocRunPos(painting.TextStyle style, string text, int offset, int size, float textScaleFactor = 1.0f) {
-            this.allocRunPos(TextStyle.applyStyle(null, style, textScaleFactor), text, offset, size);
-        }
-        
-        internal void allocRunPos(TextStyle style, string text, int offset, int size) {
-            this._style = style;
-            this._text = text;
-            this._textOffset = offset;
-            this._size = size;
-            if (this.positions == null || this.positions.Length < size) {
-                this.positions = new Vector2d[size];
-            }
-        }
-
-        public void setBounds(Rect bounds) {
-            this._bounds = bounds;
-        }
-
-        public TextBlob make() {
-            var result = new TextBlob(this._text, this._textOffset,
-                this._size, this.positions, this._bounds, this._style);
-            this.positions = null;
-            return result;
-        }
-    }
+﻿namespace Unity.UIWidgets.ui {
+    public struct TextBlob {
+        internal TextBlob(string text, int textOffset, int textSize, float[] positions,
+            UnityEngine.Rect bounds, TextStyle style) {
+            this.instanceId = ++_nextInstanceId;
+            this._positions = positions;
+            this.text = text;
+            this.textOffset = textOffset;
+            this.textSize = textSize;
+            this.style = style;
+            this._bounds = bounds;
+            this._boundsInText = null;
+        }
+
+        public Rect boundsInText {
+            get {
+                if (this._boundsInText == null) {
+                    var pos = this.getPosition(0);
+                    this._boundsInText = Rect.fromLTWH(this._bounds.xMin + pos, this._bounds.yMin,
+                        this._bounds.width, this._bounds.height);
+                }
+
+                return this._boundsInText;
+            }
+        }
+
+        public Rect shiftedBoundsInText(Offset offset) {
+            var pos = this.getPosition(0);
+            return Rect.fromLTWH(this._bounds.xMin + pos + offset.dx, this._bounds.yMin + offset.dy,
+                this._bounds.width, this._bounds.height);
+        }
+
+        public float getPosition(int i) {
+            return this._positions[this.textOffset + i];
+        }
+
+        static long _nextInstanceId;
+        internal readonly long instanceId;
+        internal readonly string text;
+        internal readonly int textOffset;
+        internal readonly int textSize;
+        internal readonly TextStyle style;
+        readonly UnityEngine.Rect _bounds; // bounds with positions[start] as origin       
+        readonly float[] _positions;

```

Comment:
In my knowledge, float[] will be allocated in heap, making the whole textblob to be a heap-allocated thing. If it is true, then making textblob a struct will not help to reduce gc that much ?

Reply:
The _positions[] here is allocated at the beginning of layout once and for all, and shared in all of TextBlobs, this array is one of the things that are attached to Paragraph and hard to eliminate. In my understanding, in C# the struct containing the reference will still be allocated on stack, only what is referenced is allocated on heap. Please point it out if I am wrong about this. In deep profile, no GC.Alloc happens any more accompanied by TextBlob.constructor, so I tend to believe this.